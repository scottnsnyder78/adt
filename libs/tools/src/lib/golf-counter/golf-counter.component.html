<div data-testid="counter-component">
  <button
    [disabled]="(current$ | async) === 0"
    data-testid="decrement-button"
    class="btn btn-primary"
    (click)="decrement()"
  >
    -
  </button>
  <span data-testid="current">{{ current$ | async }}</span>
  <button
    data-testid="increment-button"
    class="btn btn-primary"
    (click)="increment()"
  >
    +
  </button>
</div>
