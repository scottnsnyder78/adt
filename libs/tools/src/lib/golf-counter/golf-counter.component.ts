import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'adt-golf-counter',
  templateUrl: './golf-counter.component.html',
  styleUrls: ['./golf-counter.component.css'],
})
export class GolfCounterComponent implements OnInit {
  constructor() {}

  ngOnInit(): void {}
}
