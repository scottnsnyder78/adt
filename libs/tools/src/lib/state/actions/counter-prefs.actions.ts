import { createActionGroup, props, emptyProps } from '@ngrx/store';

export const CounterPrefsEvents = createActionGroup({
  source: 'Tools CounterPrefs Events',
  events: {
    bySet: props<{ by: number }>(),
  },
});

export const CounterPrefsCommands = createActionGroup({
  source: 'Tools CounterPrefs Commands',
  events: {},
});

export const CounterPrefsDocuments = createActionGroup({
  source: 'Tools CounterPrefs Documents',
  events: {},
});
