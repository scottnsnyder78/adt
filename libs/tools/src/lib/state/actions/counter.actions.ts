import { createActionGroup, props, emptyProps } from '@ngrx/store';

export const CounterEvents = createActionGroup({
  source: 'Tools Counter Events',
  events: {
    increment: emptyProps(),
    decrement: emptyProps(),
  },
});

export const CounterCommands = createActionGroup({
  source: 'Tools Counter Commands',
  events: {},
});

export const CounterDocuments = createActionGroup({
  source: 'Tools Counter Documents',
  events: {},
});
