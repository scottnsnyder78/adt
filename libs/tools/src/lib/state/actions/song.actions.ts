import { createActionGroup, props, emptyProps } from '@ngrx/store';
import { SongEntity } from '../reducers/songs.reducer';

export const SongsEvents = createActionGroup({
  source: 'List Songs Events',
  events: {
    error: props<{ message: string }>(),
  },
});

export const SongsCommands = createActionGroup({
  source: 'List Songs Commands',
  events: {
    load: emptyProps(),
  },
});

export const SongsDocuments = createActionGroup({
  source: 'List Songs Documents',
  events: {
    songs: props<{ payload: SongEntity[] }>(),
  },
});
