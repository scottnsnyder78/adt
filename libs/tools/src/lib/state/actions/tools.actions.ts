import { createActionGroup, props, emptyProps } from '@ngrx/store';

export const FeatureEvents = createActionGroup({
  source: 'Tools Feature Events',
  events: {
    entered: emptyProps(),
  },
});

export const FeatureCommands = createActionGroup({
  source: 'Tools Feature Commands',
  events: {},
});

export const FeatureDocuments = createActionGroup({
  source: 'Tools Feature Documents',
  events: {},
});
