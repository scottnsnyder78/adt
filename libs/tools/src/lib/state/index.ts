import {
  ActionReducerMap,
  createFeatureSelector,
  createSelector,
} from '@ngrx/store';
import * as fromCounter from './reducers/counter.reducer';
import * as fromSongs from './reducers/songs.reducer';
export const FEATURE_NAME = 'tools';

export interface ToolsState {
  counter: fromCounter.CounterState;
  songs: fromSongs.SongListState;
}

export const reducers: ActionReducerMap<ToolsState> = {
  counter: fromCounter.reducer,
  songs: fromSongs.reducer,
};

// 1. Feature Selector
const selectFeature = createFeatureSelector<ToolsState>(FEATURE_NAME);

// 2. Selector per "branch" of the feature.
const selectCounterBranch = createSelector(selectFeature, (f) => f.counter);
const selectSongsBranch = createSelector(selectFeature, (f) => f.songs);
// 3. Any helpers (optional)

const { selectAll: selectAllSongsArray } =
  fromSongs.adapter.getSelectors(selectSongsBranch);
// 4. What our component(s) need.

export const selectSongsList = selectAllSongsArray;

export const selectSongListLoaded = createSelector(
  selectSongsBranch,
  (b) => b.loaded,
);

export const selectCurrentCount = createSelector(
  selectCounterBranch,
  (b) => b.current,
);

export const selectCountingBy = createSelector(
  selectCounterBranch,
  (b) => b.by,
);

export const selectSongListErrored = createSelector(
  selectSongsBranch,
  (b) => b.errored,
);
