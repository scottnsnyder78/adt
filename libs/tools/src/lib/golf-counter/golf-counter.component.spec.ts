import { Spectator, createComponentFactory } from '@ngneat/spectator/jest';
import { GolfCounterComponent } from './golf-counter.component';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import * as fromGolfCounterReducer from '../state/reducers/counter.reducer';
import * as rootState from '../state';
import { CounterEvents } from '../state/actions/counter.actions';
describe('GolfCounterComponent', () => {
  let spectator: Spectator<GolfCounterComponent>;
  let store: MockStore;
  const initialState = {
    counter: {
      count: fromGolfCounterReducer.reducer,
    },
  };

  const createComponent = createComponentFactory({
    component: GolfCounterComponent,
    providers: [
      provideMockStore({
        initialState,
        selectors: [
          {
            selector: rootState.selectCurrentCount,
            value: 444,
          },
        ],
      }),
    ],
  });

  beforeEach(() => {
    spectator = createComponent();
    store = spectator.inject(MockStore);
    spectator.detectChanges();
  });

  it('Initial State', () => {
    const component = spectator.query('[data-testid="counter-component"]');
    expect(component).toMatchSnapshot();
  });

  it('can increment', () => {
    const incrementButton = spectator.query('[data-testid="increment-button"]');
    if (!incrementButton) {
      fail('No Increment button found');
    } else {
      jest.spyOn(store, 'dispatch');
      spectator.click(incrementButton);
      expect(store.dispatch).toHaveBeenCalledWith(CounterEvents.increment());
    }
  });
  it('demo of just a snapshot of a span', () => {
    const component = spectator.query('[data-testid="current"]');
    expect(component).toMatchSnapshot();
  });
});
