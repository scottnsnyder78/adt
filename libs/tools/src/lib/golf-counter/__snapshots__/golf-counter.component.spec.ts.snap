// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GolfCounterComponent Initial State 1`] = `
<div
  data-testid="counter-component"
>
  <button
    class="btn btn-primary"
    data-testid="decrement-button"
  >
     Decrease 
  </button>
  <span
    data-testid="current"
  >
    444
  </span>
  <button
    class="btn btn-primary"
    data-testid="increment-button"
  >
     Increase 
  </button>
</div>
`;

exports[`GolfCounterComponent demo of just a snapshot of a span 1`] = `
<span
  data-testid="current"
>
  444
</span>
`;
